### PR TITLE
mold: wrap so it can find system libraries

### PR DESCRIPTION
**Pull request by dpc**
## Description of changes

Add a wrapped version as `mold` as `mold-wrapped`.

Currently `mold` mostly works, but the moment users start to need to link with system libraries, they hit issues.

github:

Re #242816

and on user forums, e.g.:

https://discourse.nixos.org/t/using-mold-as-linker-prevents-libraries-from-being-found

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/mold/default.nix
Comment (dpc) [inline pkgs/development/tools/mold/default.nix:32]: Does this even work in `symlinkJoin`? :D

**Comment by Artturin** [comment]
I added `useMoldLinker` `stdenv` adapter to use mold https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/adapters.nix

**Comment by dpc** [comment]
> I added `useMoldLinker` `stdenv` adapter to use mold [299a7bd](https://github.com/NixOS/nixpkgs/commit/299a7bd35e552fe6e444096658d7e9ebafb038af)

I'm not very versed with `stdenv`. I don't think it's very helpful for e.g. Rust users, that will just change the linker in `.cargo/config.toml` and expect it to work, just like on other Linuxes. Is there something I'm missing here?

File: pkgs/top-level/all-packages.nix
Comment (Artturin) [inline pkgs/top-level/all-packages.nix:19849]: Can be simplified to just

```suggestion
  mold = wrapBintoolsWith { bintools = mold-unwrapped; };
  
  mold-unwrapped = callPackage ../development/tools/mold-unwrapped { };
```

**Comment by Artturin** [comment]
Im worried about having multiple bintools in an environment because AFAIK there hasn't been a case like that before and so all the setup assume one bintools which contains all the utils

**Comment by dpc** [comment]
> Im worried about having multiple bintools in an environment because AFAIK there hasn't been a case like that before and so all the setup assume one bintools which contains all the utils.

I don't understand the details here, but ...

> I think we should start by having a explicit `mold-wrapped` and leaving mold unwrapped

... that's good enough for me. Better than copy&pasting some script between my flakes, and a simple solution for affected users, I guess.

**Comment by dpc** [comment]
@Artturin Done, hopefully I combined the both suggestions correctly.

**Comment by Artturin** [comment]
Remember to look at the CI

drop or fixup and reword the unnecessary commit

**Comment by Artturin** [comment]
Before merge let's create tests for wrapped mold and the mold adapter  in pkgs/test/

I will do it but you can take a shot at it too if you want

File: maintainers/maintainer-list.nix
Comment (dpc) [reply maintainers/maintainer-list.nix:4552]: > Looks like you also forgot to add yourself as a maintainer of the package.

Oh, I left it in by mistake because I'm waiting for https://github.com/NixOS/nixpkgs/pull/251689 and didn't want to create a commit confict ...

But I actually did not intend to add myself to maintainer of mold itself. If being an interested user enough, I can, but it wasn't my initial intention. :D

**Comment by dpc** [comment]
> I will do it but you can take a shot at it too if you want

Please go ahead if you can. I'll be happy to review and learn something, but I have no idea how to start, and not a lot of time ATM.

**Comment by Artturin** [comment]
Tests added

**Review by Artturin — APPROVED** [review]

**Comment by nh2** [comment]
For reference, I highlighted why `mold-wrapped` is better in many cases than `stdenvAdapters.useMoldLinker` [here](https://github.com/NixOS/nixpkgs/issues/215260#issuecomment-3067161637).

**Pull request merged**
